src/khepri_projection.erl

```diff
@@ -0,0 +1,296 @@
+%% file, You can obtain one at https://mozilla.org/MPL/2.0/.
+%%
+%% Copyright © 2022 VMware, Inc. or its affiliates.  All rights reserved.
+%%
+
+%% @doc Khepri projections
+%%
+%% Projections build a replicated ETS table using nodes from the store which
+%% match a {@link khepri_path:pattern()}. When a node matching a projection's
+%% pattern is changed in the store, the node is passed through the projection's
+%% {@link projection_fun()} to create record(s). These records are then stored
+%% in the projection's ETS table for all members in a Khepri cluster.
+%%
+%% Projections provide a way to query the store as fast as possible and are
+%% appropriate for lookups which require low latency and/or high throughput.
+%% Projection tables contain all records matching the pattern, though, so the
+%% memory footprint of a projection table grows with the number of nodes in
+%% the store matching the pattern.
+%%
+%% Projection ETS tables are owned by the Khepri cluster and are deleted when
+%% the cluster stops.
+%%
+%% Updates to projection tables are immediately consistent for the member of
+%% the cluster on which the change to the store is made and the leader member
+%% but are eventually consistent for all other followers.
+
+-module(khepri_projection).
+
+-include_lib("kernel/include/logger.hrl").
+
+-include("khepri_machine.hrl").
+
+-export([new/2, new/3, name/1]).
+
+ %% For internal use only
+-export([init/1, trigger/4]).
+
+-export_type([projection/0,
+              projection_fun/0,
+              options/0]).
+
+-define(PAYLOAD_OR(Payload, Default),
+        (if Payload =:= ?NO_PAYLOAD -> Default; true -> Payload end)).
+
+-record(?MODULE,
+        {%% The name of the projection. This atom should be unique to a node
+         %% as it will become the name of the ETS table.
+         name :: atom(),
+         %% A function compiled by `khepri_fun' which "projects" entries from
+         %% a Khepri store into records stored in a "projected" ETS table.
+         project_fun :: {simple | extended, khepri_fun:standalone_fun()},
+         %% Options passed to `ets:new/2'
+         ets_options :: [atom() | tuple()]}).
+
+-opaque projection() :: #?MODULE{}.
+
+-type simple_projection_fun() ::
+fun((Path :: khepri_path:native_path(),
+     Payload :: term() | undefined) -> Record :: tuple()).
+%% A simple projection function.
+%%
+%% Simple projection functions only take the path and data for a node in the
+%% store. The record produced by the function is used to create and delete
+%% objects in the ETS table.
+%%
+%% `Payload' is `undefined' if there is no payload for the node.
+%%
+%% This function is compiled the same way as a transaction function: all
+%% side-effects are not allowed.
+
+-type extended_projection_fun() ::
+fun((Table :: ets:tid(),
+     Path :: khepri_path:native_path(),
+     OldPayload :: term() | undefined,
+     NewPayload :: term() | undefined) -> any()).
```

Comment:
I find it a bit confusing that the payload goes through several forms between the state machine and the final call to the projection function.

What do you think of using a node properties map (see `khepri_machine:gather_node_props()`) to transport the old and new payloads all to the way to the call of the projection function? This way, you can distinguish between a stored `undefined` atom and the absence of data payload.